radix-engine/src/model/resource_manager.rs

```diff
@@ -173,39 +169,77 @@ impl ResourceManager {
             (UpdateNonFungibleData, (DenyAll, LOCKED)),
         ] {
             let entry = auth.remove(&auth_entry_key).unwrap_or(default);
-            authorization.insert(auth_entry_key, MethodEntry::new(entry));
+            authorization.insert(auth_entry_key, MethodAccessRule::new(entry));
         }
 
         let resource_manager = Self {
             resource_type,
             metadata,
             method_table,
+            vault_method_table,
             authorization,
             total_supply: 0.into(),
         };
 
         Ok(resource_manager)
     }
 
-    pub fn get_auth(&self, method_name: &str, args: &[ScryptoValue]) -> &MethodAuthorization {
-        if method_name.eq("method_auth") {
-            let method: ResourceMethod = match scrypto_decode(&args[0].raw) {
-                Ok(r) => r,
-                Err(_) => return &MethodAuthorization::Unsupported,
-            };
-
-            match self.authorization.get(&method) {
-                None => &MethodAuthorization::Unsupported,
-                Some(entry) => {
-                    let auth_args = args.split_at(1).1;
-                    entry.get_update_auth(auth_args)
+    pub fn get_vault_auth(&self, arg: &ScryptoValue) -> &MethodAuthorization {
+        let method: VaultMethod = match scrypto_decode(&arg.raw) {
+            Ok(m) => m,
+            Err(_) => return &MethodAuthorization::Unsupported,
+        };
+
+        match self.vault_method_table.get(method.name()) {
+            None => &MethodAuthorization::Unsupported,
+            Some(None) => &MethodAuthorization::AllowAll,
+            Some(Some(method)) => self.authorization.get(method).unwrap().get_method_auth(),
+        }
```

Comment:
Why don't we store `MethodAuthorization` to the vault method? Looks like unnecessary indirection.

Reply:
As all vaults from a resource get their authorization from a single auth source (required so that resource manager can update this easily). Having said that, a bunch of refactors still needed to be done to get this in a better state (e.g. moving the auth stuff outside of resource manager).

Reply:
Not sure if I get the point. 

Saw a lot of redundant matching arms like this.
```rust
None => &MethodAuthorization::Unsupported,
Some(None) => &MethodAuthorization::AllowAll,
Some(Some(method)) => self.authorization.get(method).unwrap().get_method_auth(),
```

Originally, I think I didn't understand why we need an `Option` wrapping `MethodAuthorization`, as it seems we're always assigning a value to all methods and this is inconsistent with component auth. 

But digging into this a bit more, looks like we're only using it as a shortcut for "if the method auth is none, there is no update/lock_auth method", which related to my other point about polymorphism.  This worsens the situation as we're essentially introducing an ABI per resource configuration.